### PR TITLE
Improve tuple & array inter-assignability

### LINKED ISSUE #13167: [Spec Deviation] A tuple where all the member-type-descriptors are the same should be equal to an  array-type-descriptor
**Description:**
As per the spec (v0.990:2019-01-16),
"Note that a tuple type where all the member-type-descriptors are the same is equivalent to a an array-type-descriptor with a length."
```ballerina
@test:Config {}
function testTupleAndArrayTypeDescriptorBroken() {
    int[3] intArray = [1, 2, 3];
    [int, int, int] b = intArray; // This should work

    [string, string, string] stringT = ["a", "b", "c"];
    string[3] stringA = stringT; // This should work
}
```

**Affected Versions:**
0.990.2

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$subject, so that it'll allow below samples;

Fixes #13167

## Samples
```ballerina
    string[3] a= ["a", "b", "c"];
    [string, string, string] b = a;

    [string, string, string] c = ["a", "b", "c"];
    string[3] d = c;
```

## Remarks
https://github.com/ballerina-platform/ballerina-lang/issues/13167

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -734,6 +738,15 @@ private boolean isTupleTypeAssignable(BType source, BType target, List<TypePair>
 
     private boolean isTupleTypeAssignableToArrayType(BTupleType source, BArrayType target,
                                                      List<TypePair> unresolvedTypes) {
+
+        if (source.restType != null && target.state != BArrayState.UNSEALED) {
+            return false;
+        }
+
+        if (target.state != BArrayState.UNSEALED && source.tupleTypes.size() != target.size) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Shall we change L742-L748 to do the `target.state != BArrayState.UNSEALED` check once?

```ballerina
if (target.state != BArrayState.UNSEALED && (source.restType != null || source.tupleTypes.size() != target.size)) {
    return false;
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -742,6 +755,24 @@ private boolean isTupleTypeAssignableToArrayType(BTupleType source, BArrayType t
                 .allMatch(tupleElemType -> isAssignable(tupleElemType, target.eType, unresolvedTypes));
     }
 
+    private boolean isArrayTypeAssignableToTupleType(BArrayType source, BTupleType target,
+                                                     List<TypePair> unresolvedTypes) {
+        if (target.restType != null && source.state != BArrayState.UNSEALED) {
+            return false;
+        }
+
+        if (source.state != BArrayState.UNSEALED && target.tupleTypes.size() != source.size) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: We can merge this too right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 1 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed in dbeae05

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 2 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed in dbeae05

**Pull request merged**